controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -201,9 +206,25 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
                              .thenAccept(isDone::set), executor);
     }
 
-    private CompletableFuture<Boolean> isUpdated(String scope, String stream, StreamConfiguration newConfig, OperationContext context) {
-        return streamMetadataStore.getConfigurationRecord(scope, stream, context, executor)
-                .thenApply(configProperty -> !configProperty.getObject().isUpdating() || !configProperty.getObject().getStreamConfiguration().equals(newConfig));
+    @VisibleForTesting
+    CompletableFuture<Boolean> isUpdated(String scope, String stream, StreamConfiguration newConfig, OperationContext context) {
+        CompletableFuture<State> stateFuture = streamMetadataStore.getState(scope, stream, true, context, executor);
+        CompletableFuture<StreamConfigurationRecord> configPropertyFuture
+                = streamMetadataStore.getConfigurationRecord(scope, stream, context, executor).thenApply(VersionedMetadata::getObject);
+        return CompletableFuture.allOf(stateFuture, configPropertyFuture)
+                                .thenApply(v -> {
+                                    State state = stateFuture.join();
+                                    StreamConfigurationRecord configProperty = configPropertyFuture.join();
+
+                                    // if property is updating and doesnt match our request, its a subsequent update
+                                    if (configProperty.isUpdating()) {
+                                        return !configProperty.getStreamConfiguration().equals(newConfig);
+                                    } else {
+                                        // if update-barrier is not updating, then update is complete if property matches our expectation 
+                                        // and state is not updating 
+                                        return !(configProperty.getStreamConfiguration().equals(newConfig) && state.equals(State.UPDATING));
```

Comment:
The new condition in the description:

```
ConfigProp.config == newConfig && !configProp.isUpdating ==> inProgress
&& state != TRUNCATING // if truncation record is updated but scale is not reset ==> updating
```

does not seem to match the one here. In particular, if `ConfigProp.config != newConfig` (configs differ), then we end up evaluating to true, which I don't understand whether it is correct. The condition in the code is equivalent to:

```
ConfigProp.config != newConfig || !configProp.isUpdating || state != UPDATING ==> updated
```

or 

```
ConfigProp.config == newConfig && configProp.isUpdating && state == UPDATING ==> is updating
```

Reply:
truncation and update have symmetrical checks. the comment above applies here too.

When we submit a new update, we set the ConfigurationRecord to updating with new configuration in it.

So first condition to check is `configurationRecord.isUpdating`

case 1 : configurationRecord.isUpdating = true
that means there is an ongoing update request.
If ongoing request doesnt have the same configuration as that we requested, then our request has definitely completed else our request is definitely incomplete.

case 2: configurationRecord.isUpdating = false
Otherwise ongoing update work is complete. Stream State may or may not have been reset.
So, if configurationRecord.config matches our request and state is UPDATING, then we completed the update but are yet to reset the state. we will return InProgress.

Reply:
> ConfigProp.config == newConfig && configProp.isUpdating && state == UPDATING ==> is updating
Note: this is an incomplete condition. Update to `configProp` and `state` are not atomic. 
So `configProp.isUpdating` maybe false but state may be updating. 
We want to return InProgress in such cases. That is what this PR is doing, earlier we would return success by only checking `configProp` and not include `state`.

Reply:
> In particular, if ConfigProp.config != newConfig (configs differ), then we end up evaluating to true, which I don't understand whether it is correct. The condition in the code is equivalent to:

this is because there can only be one outstanding update request. 
So if the outstanding update request does not match the one we submitted then our requested udpate is definitely done

Reply:
oops sorry again.. 
``` configProp.config == ourConfig && state!= UPDATING // if config record is updated but state is not reset   
 ==> InProgress
 else ==> done
```
there is problem in the description.. the first condition has `state == UPDATING`.. i wrote it correctly in words.. but not in the pseudo code..